File: Vantage/Models/Core/MarketData.swift
import Foundation
import SwiftData

/// A point-in-time quote observation.
///
/// Append-only. Each fetch inserts a new row rather than overwriting the last,
/// which is what makes "what did this look like when I last opened the app"
/// answerable without a separate snapshot mechanism.
@Model
final class QuoteObservation {
    var security: Security?
    /// When we recorded this. Distinct from `quoteTime`, which is the exchange's
    /// timestamp — they differ, and the gap matters for stale-data detection.
    var observedAt: Date
    var quoteTime: Date?

    var last: Double
    var open: Double?
    var high: Double?
    var low: Double?
    var previousClose: Double?
    var volume: Double?

    var providerRaw: String

    init(
        security: Security? = nil,
        observedAt: Date = .now,
        quoteTime: Date? = nil,
        last: Double,
        open: Double? = nil,
        high: Double? = nil,
        low: Double? = nil,
        previousClose: Double? = nil,
        volume: Double? = nil,
        provider: DataProviderID = .finnhub
    ) {
        self.security = security
        self.observedAt = observedAt
        self.quoteTime = quoteTime
        self.last = last
        self.open = open
        self.high = high
        self.low = low
        self.previousClose = previousClose
        self.volume = volume
        self.providerRaw = provider.rawValue
    }

    var provider: DataProviderID { DataProviderID(rawValue: providerRaw) ?? .finnhub }

    /// Absolute and percentage change against the prior session's close.
    /// Nil when the provider didn't supply a previous close — the spec forbids
    /// inventing a value to fill the gap.
    var change: Double? {
        guard let previousClose else { return nil }
        return last - previousClose
    }

    var changePercent: Double? {
        guard let previousClose, previousClose != 0 else { return nil }
        return (last - previousClose) / previousClose * 100
    }
}

/// One OHLCV bar. Bars are immutable once written for a given
/// (security, date, resolution); a re-fetch that disagrees indicates a
/// provider restatement and is recorded as a new row with a later `observedAt`.
@Model
final class PriceBar {
    var security: Security?
    /// The session or interval this bar describes.
    var date: Date
    var resolutionRaw: String
    var observedAt: Date

    var open: Double
    var high: Double
    var low: Double
    var close: Double
    var volume: Double?
    /// Split/dividend-adjusted close where the provider supplies one. Returns
    /// computed over long windows must use this, not raw close.
    var adjustedClose: Double?

    init(
        security: Security? = nil,
        date: Date,
        resolution: BarResolution,
        observedAt: Date = .now,
        open: Double,
        high: Double,
        low: Double,
        close: Double,
        volume: Double? = nil,
        adjustedClose: Double? = nil
    ) {
        self.security = security
        self.date = date
        self.resolutionRaw = resolution.rawValue
        self.observedAt = observedAt
        self.open = open
        self.high = high
        self.low = low
        self.close = close
        self.volume = volume
        self.adjustedClose = adjustedClose
    }

    var resolution: BarResolution { BarResolution(rawValue: resolutionRaw) ?? .daily }

    /// Prefers the adjusted series when available, since that is what any
    /// return or moving-average calculation should be built on.
    var analysisClose: Double { adjustedClose ?? close }
}

enum BarResolution: String, Codable, Sendable, CaseIterable {
    case oneMinute = "1"
    case fiveMinute = "5"
    case fifteenMinute = "15"
    case hourly = "60"
    case daily = "D"
    case weekly = "W"
    case monthly = "M"
}

/// The chart ranges from Section 5.
enum ChartRange: String, CaseIterable, Identifiable, Sendable {
    case oneDay = "1D"
    case fiveDay = "5D"
    case oneMonth = "1M"
    case threeMonth = "3M"
    case sixMonth = "6M"
    case oneYear = "1Y"
    case fiveYear = "5Y"

    var id: String { rawValue }

    /// The bar resolution that gives a readable chart without over-fetching.
    var resolution: BarResolution {
        switch self {
        case .oneDay: .fiveMinute
        case .fiveDay: .fifteenMinute
        case .oneMonth, .threeMonth, .sixMonth, .oneYear: .daily
        case .fiveYear: .weekly
        }
    }

    /// Whether this range needs a series finer than daily — and therefore a
    /// different vendor, with a different share of the tape behind it.
    var usesIntraday: Bool { !resolution.isDailyOrCoarser }

    var dateInterval: DateComponents {
        switch self {
        case .oneDay: DateComponents(day: -1)
        case .fiveDay: DateComponents(day: -5)
        case .oneMonth: DateComponents(month: -1)
        case .threeMonth: DateComponents(month: -3)
        case .sixMonth: DateComponents(month: -6)
        case .oneYear: DateComponents(year: -1)
        case .fiveYear: DateComponents(year: -5)
        }
    }

    func startDate(from end: Date = .now, calendar: Calendar = .current) -> Date {
        calendar.date(byAdding: dateInterval, to: end) ?? end
    }

    /// How many trading sessions an intraday range shows, if it is one.
    ///
    /// Sessions, not calendar days: five days back from a Monday afternoon
    /// reaches the previous Wednesday, so "5D" drew three sessions and called
    /// them five.
    var intradaySessions: Int? {
        switch self {
        case .oneDay: 1
        case .fiveDay: 5
        default: nil
        }
    }

    /// How far back to ask an intraday vendor for bars.
    ///
    /// Deliberately wider than the range itself, because the count that
    /// matters is sessions and the calendar does not know which days traded.
    /// The surplus is trimmed once the bars are in hand.
    func intradayFetchStart(from end: Date = .now, calendar: Calendar = .current) -> Date {
        guard let sessions = intradaySessions else {
            return startDate(from: end, calendar: calendar)
        }
        let days = sessions == 1 ? -5 : -(sessions * 2 + 4)
        return calendar.date(byAdding: DateComponents(day: days), to: end) ?? end
    }
}


/// Whether a chart can be drawn, and what to say when it cannot.
///
/// Exists because "no bars" and "still loading" were previously the same
/// state on screen, which is how 1D and 5D came to show a spinner that never
/// resolved for a range that had nothing to load.
enum ChartAvailability: Equatable, Sendable {
    case ready
    case loading
    case unavailable(String)
}


/// One point on an intraday chart, placed by position rather than by clock.
///
/// A wall-clock axis gives 17 of every 24 hours to time that did not trade:
/// on 5D three sessions occupied about a fifth of the width and the rest was
/// blank. Placing bars end to end spends the whole axis on trading, and the
/// axis labels below say which session each stretch belongs to.
struct ChartPoint: Identifiable {
    let id: Int
    let date: Date
    let close: Double
    /// Start of the trading day this point belongs to, in New York.
    let session: Date

    var position: Double { Double(id) }
}


/// A labelled position on the intraday chart's x-axis.
struct ChartAxisTick: Identifiable {
    let id: Int
    let label: String

    var position: Double { Double(id) }
}

File: core/src/commonMain/kotlin/com/tylerabitbol/libra/models/core/MarketData.kt
package com.tylerabitbol.libra.models.core

import kotlin.time.Clock
import kotlin.time.Instant
import kotlinx.datetime.DatePeriod
import kotlinx.datetime.TimeZone
import kotlinx.datetime.minus

enum class BarResolution(val raw: String) {
    OneMinute("1"),
    FiveMinute("5"),
    FifteenMinute("15"),
    Hourly("60"),
    Daily("D"),
    Weekly("W"),
    Monthly("M");

    /** Daily or coarser bars come from a different vendor than intraday ones. */
    val isDailyOrCoarser: Boolean get() = this == Daily || this == Weekly || this == Monthly

    companion object {
        fun fromRaw(raw: String?): BarResolution? = entries.firstOrNull { it.raw == raw }
    }
}

/** The chart ranges from Section 5. */
enum class ChartRange(val raw: String) {
    OneDay("1D"),
    FiveDay("5D"),
    OneMonth("1M"),
    ThreeMonth("3M"),
    SixMonth("6M"),
    OneYear("1Y"),
    FiveYear("5Y");

    val id: String get() = raw

    /** The bar resolution that gives a readable chart without over-fetching. */
    val resolution: BarResolution
        get() = when (this) {
            OneDay -> BarResolution.FiveMinute
            FiveDay -> BarResolution.FifteenMinute
            OneMonth, ThreeMonth, SixMonth, OneYear -> BarResolution.Daily
            FiveYear -> BarResolution.Weekly
        }

    /**
     * Whether this range needs a series finer than daily — and therefore a
     * different vendor, with a different share of the tape behind it.
     */
    val usesIntraday: Boolean get() = !resolution.isDailyOrCoarser

    val datePeriod: DatePeriod
        get() = when (this) {
            OneDay -> DatePeriod(days = 1)
            FiveDay -> DatePeriod(days = 5)
            OneMonth -> DatePeriod(months = 1)
            ThreeMonth -> DatePeriod(months = 3)
            SixMonth -> DatePeriod(months = 6)
            OneYear -> DatePeriod(years = 1)
            FiveYear -> DatePeriod(years = 5)
        }

    fun startDate(
        end: Instant = Clock.System.now(),
        zone: TimeZone = TimeZone.currentSystemDefault()
    ): Instant = end.minus(datePeriod, zone)

    /**
     * How many trading sessions an intraday range shows, if it is one.
     *
     * Sessions, not calendar days: five days back from a Monday afternoon
     * reaches the previous Wednesday, so "5D" drew three sessions and called
     * them five.
     */
    val intradaySessions: Int?
        get() = when (this) {
            OneDay -> 1
            FiveDay -> 5
            else -> null
        }

    /**
     * How far back to ask an intraday vendor for bars.
     *
     * Deliberately wider than the range itself, because the count that
     * matters is sessions and the calendar does not know which days traded.
     * The surplus is trimmed once the bars are in hand.
     */
    fun intradayFetchStart(
        end: Instant = Clock.System.now(),
        zone: TimeZone = TimeZone.currentSystemDefault()
    ): Instant {
        val sessions = intradaySessions ?: return startDate(end, zone)
        val days = if (sessions == 1) 5 else sessions * 2 + 4
        return end.minus(DatePeriod(days = days), zone)
    }

    companion object {
        fun fromRaw(raw: String?): ChartRange? = entries.firstOrNull { it.raw == raw }
    }
}

/**
 * Whether a chart can be drawn, and what to say when it cannot.
 *
 * Exists because "no bars" and "still loading" were previously the same
 * state on screen, which is how 1D and 5D came to show a spinner that never
 * resolved for a range that had nothing to load.
 */
sealed interface ChartAvailability {
    data object Ready : ChartAvailability
    data object Loading : ChartAvailability
    data class Unavailable(val reason: String) : ChartAvailability
}

/**
 * One point on an intraday chart, placed by position rather than by clock.
 *
 * A wall-clock axis gives 17 of every 24 hours to time that did not trade:
 * on 5D three sessions occupied about a fifth of the width and the rest was
 * blank. Placing bars end to end spends the whole axis on trading, and the
 * axis labels below say which session each stretch belongs to.
 */
data class ChartPoint(
    val id: Int,
    val date: Instant,
    val close: Double,
    /** Start of the trading day this point belongs to, in New York. */
    val session: Instant
) {
    val position: Double get() = id.toDouble()
}

/**
 * A run of the intraday line that is drawn as one stroke.
 *
 * Sessions and the moves between them are drawn differently but belong to
 * one line. Splitting the chart into disconnected per-session lines was a
 * defence against a wall-clock axis, where the weekend spanned a fifth of
 * the width; on a positional axis the space between a close and the next
 * open is a single bar, and a stroke that narrow reads as the jump it is.
 */
data class ChartSegment(
    /** Also the chart's series key, which is why it is a `String` for both kinds. */
    val id: String,
    val kind: Kind,
    val points: List<ChartPoint>
) {
    enum class Kind {
        /** Bars that traded, consecutively, within one session. */
        Traded,

        /**
         * The step from one session's last close to the next session's open.
         * Exactly two points, one position apart.
         */
        Overnight
    }
}

/** A labelled position on the intraday chart's x-axis. */
data class ChartAxisTick(val id: Int, val label: String) {
    val position: Double get() = id.toDouble()
}
